### PR TITLE
Batch update with singleton/sharding fixes

### PR DESCRIPTION
**Pull request by zbynek001**
Mostly singleton & sharding fixes, plus some dependencies
Will split it into smaller parts if needed

Included:
https://github.com/akka/akka/pull/25854
https://github.com/akka/akka/issues/25274
https://github.com/akka/akka/pull/24442
https://github.com/akka/akka/issues/25639
https://github.com/akka/akka/issues/26336
https://github.com/akka/akka/issues/26793
https://github.com/akka/akka/pull/25648
https://github.com/akka/akka/pull/26061
https://github.com/akka/akka/issues/26012
https://github.com/akka/akka/issues/26214
https://github.com/akka/akka/issues/26451
https://github.com/akka/akka/pull/26819
& persistent shard passivate-idle fix

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
Let me know when this is ready to review - I'm working on those serialization fixes over here: https://github.com/akkadotnet/akka.net/pull/3744

**Comment by zbynek001** [comment]
I think this is mostly ready, there is one open topic with persistent sharding passivate-idle

**Comment by Aaronontheweb** [comment]
@zbynek001 looks like there might be some conflicts with the most recent changes in `dev` I pushed earlier today.

**Comment by Aaronontheweb** [comment]
Looks like this is already covered inside https://github.com/akkadotnet/akka.net/pull/3801

**Comment by Aaronontheweb** [comment]
Compilation issue on Linux:

```
Singleton/ClusterSingletonLeavingSpeedSpec.cs(118,18): error CS8137: Cannot define a class or member that utilizes tuples because the compiler required type 'System.Runtime.CompilerServices.TupleElementNamesAttribute' cannot be found. Are you missing a reference? [/home/vsts/work/1/s/src/contrib/cluster/Akka.Cluster.Tools.Tests/Akka.Cluster.Tools.Tests.csproj]
Singleton/ClusterSingletonLeavingSpeedSpec.cs(118,18): error CS8356: Predefined type 'System.ValueTuple`2' is declared in multiple referenced assemblies: 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' and 'System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' [/home/vsts/work/1/s/src/contrib/cluster/Akka.Cluster.Tools.Tests/Akka.Cluster.Tools.Tests.csproj]
```

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests/CoordinatedShutdownShardingSpec.cs [hunk 1]
```diff
@@ -105,12 +105,15 @@ protected override void BeforeTermination()
         /// </summary>
         private void PingEntities()
         {
-            _region2.Tell(1, _probe2.Ref);
-            _probe2.ExpectMsg<int>(10.Seconds()).Should().Be(1);
-            _region2.Tell(2, _probe2.Ref);
-            _probe2.ExpectMsg<int>(10.Seconds()).Should().Be(2);
-            _region2.Tell(3, _probe2.Ref);
-            _probe2.ExpectMsg<int>(10.Seconds()).Should().Be(3);
+            AwaitAssert(() =>
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests/CoordinatedShutdownShardingSpec.cs]: Good change

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests/InactiveEntityPassivationSpec.cs [hunk 2]
```diff
@@ -69,57 +69,112 @@ public GotIt(string id, object msg, long when)
...
+                        return Id == other.Id;
+                    return false;
+                }
             }
         }
 
         #endregion
 
+        protected ClusterShardingSettings settings;
+        protected readonly TimeSpan smallTolerance = TimeSpan.FromMilliseconds(300);
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests/InactiveEntityPassivationSpec.cs]: Yep, saw this when I started work on #3780 - good change

File: src/contrib/cluster/Akka.Cluster.Sharding/ClusterSharding.cs [hunk 3]
```diff
@@ -1109,7 +1109,7 @@ public IActorRef ShardRegionProxy(string typeName)
             throw new ArgumentException($"Shard type [{typeName}] must be started first");
         }
 
-        private IShardAllocationStrategy DefaultShardAllocationStrategy(ClusterShardingSettings settings)
+        public IShardAllocationStrategy DefaultShardAllocationStrategy(ClusterShardingSettings settings)
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Sharding/ClusterSharding.cs]: Good change

File: src/core/Akka.Persistence/Eventsourced.cs [hunk 4]
```diff
@@ -455,6 +455,30 @@ public void DeleteMessages(long toSequenceNr)
...
+        /// range that are less than or equal to the given <see cref="SnapshotMetadata.SequenceNr"/>
+        /// (provided the <see cref="SnapshotMetadata.SequenceNr"/> is &lt;= to <see cref="Eventsourced.LastSequenceNr"/>).
+        ///
+        /// Or delete all by using `long.MaxValue` as the `toSequenceNr`
+        /// {{{ m.copy(sequenceNr = long.MaxValue) }}}
+        /// </summary>
+        /// <param name="e"></param>
+        /// <param name="keepNrOfBatches"></param>
+        /// <param name="snapshotAfter"></param>
+        internal void InternalDeleteMessagesBeforeSnapshot(SaveSnapshotSuccess e, int keepNrOfBatches, int snapshotAfter)
```
Comment (Aaronontheweb) [inline src/core/Akka.Persistence/Eventsourced.cs]: Like this - not user-friendly enough to make `public` but good enough for our internal uses of `Eventsourced`.

File: src/core/Akka.Cluster/CoordinatedShutdownLeave.cs [hunk 5]
```diff
@@ -45,7 +45,7 @@ public CoordinatedShutdownLeave()
             {
                 // MemberRemoved is needed in case it was downed instead
                 _cluster.Leave(_cluster.SelfAddress);
-                _cluster.Subscribe(Self, typeof(ClusterEvent.MemberLeft), typeof(ClusterEvent.MemberRemoved));
+                _cluster.Subscribe(Self, typeof(ClusterEvent.MemberLeft), typeof(ClusterEvent.MemberRemoved), typeof(ClusterEvent.MemberDowned));
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/CoordinatedShutdownLeave.cs]: Good catch

File: src/core/Akka.Cluster/ClusterEvent.cs [hunk 6]
```diff
@@ -885,6 +901,9 @@ private static IEnumerable<IMemberEvent> CollectMemberEvents(IEnumerable<Member>
                     case MemberStatus.Exiting:
                         yield return new MemberExited(member);
                         break;
+                    case MemberStatus.Down:
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterEvent.cs]: Also a good catch

File: src/core/Akka.Cluster/ClusterEvent.cs [hunk 7]
```diff
@@ -341,6 +341,22 @@ public MemberExited(Member member)
             }
         }
 
+        /// <summary>
+        /// Member status changed to `MemberStatus.Down` and will be removed
+        /// when all members have seen the `Down` status.
+        /// </summary>
+        public sealed class MemberDowned : MemberStatusChange
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterEvent.cs]: LGTM

File: src/contrib/cluster/Akka.Cluster.Tools/Singleton/OldestChangedBuffer.cs [hunk 8]
```diff
@@ -153,10 +153,15 @@ private bool MatchingRole(Member member)
         private void HandleInitial(ClusterEvent.CurrentClusterState state)
         {
             _membersByAge = state.Members
-                .Where(m => (m.Status == MemberStatus.Up || m.Status == MemberStatus.Leaving) && MatchingRole(m))
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Tools/Singleton/OldestChangedBuffer.cs]: Makes sense - wouldn't want to pin a `ClusterSingleton` on a node that's leaving.

File: src/contrib/cluster/Akka.Cluster.Tools.Tests/Akka.Cluster.Tools.Tests.csproj [hunk 9]
```diff
@@ -13,6 +13,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(TestSdkVersion)" />
+    <PackageReference Include="System.ValueTuple" Version="4.4.0" />
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.Cluster.Tools.Tests/Akka.Cluster.Tools.Tests.csproj]: Looks like this is what's causing the issues on Linux - let me take a look and see if I can figure out why.

**Comment by Aaronontheweb** [comment]
The `ValueTuple` compilation error appears to be caused by something that's installed on the machines running on Azure Pipelines - could be related to this: https://stackoverflow.com/a/44770334/377476

**Comment by Aaronontheweb** [comment]
Only affects Linux though.

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 10]
```diff
@@ -2172,6 +2175,10 @@ private void ShutdownSelfWhenDown()
                     SendGossipRandom(MaxGossipsBeforeShuttingDownMyself);
                     Shutdown();
                 }
+                else
+                {
+                    _selfDownCounter++;
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: Makes sense - gives the node a chance to gossip a few times prior to termination.

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 11]
```diff
@@ -2315,6 +2323,26 @@ public void LeaderActionsOnConvergence()
                 }
 
                 Publish(_latestGossip);
+                GossipExitingMembersToOldest(changedMembers.Where(i => i.Status == MemberStatus.Exiting));
+            }
+        }
+
+        /// <summary>
+        /// Gossip the Exiting change to the two oldest nodes for quick dissemination to potential Singleton nodes
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: This optimization makes sense

File: src/core/Akka.Cluster/ClusterDaemon.cs [hunk 12]
```diff
@@ -2465,6 +2493,35 @@ public bool ValidNodeForGossip(UniqueAddress node)
...
+        /// The Exiting change is gossiped to the two oldest nodes for quick dissemination to potential Singleton nodes
+        /// </summary>
+        /// <param name="latestGossip"></param>
+        /// <param name="exitingMembers"></param>
+        /// <returns></returns>
+        public static IEnumerable<Member> GossipTargetsForExitingMembers(Gossip latestGossip, IEnumerable<Member> exitingMembers)
+        {
+            if (exitingMembers.Any())
+            {
+                var roles = exitingMembers.SelectMany(m => m.Roles);
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/ClusterDaemon.cs]: Might be enumerating the `existingMembers` collection multiple times, but this doesn't appear to be an issue per the actual unit tests and MNTR results.

**Comment by Aaronontheweb** [comment]
thanks for your contribution @zbynek001

**Pull request merged**